site/source/docs/compiling/Building-Projects.rst

```diff
@@ -177,7 +177,7 @@ Emscripten Ports is a collection of useful libraries, ported to Emscripten. They
 
 You should see some notifications about SDL2 being used, and built if it wasn't previously. You can then view ``sdl2.html`` in your browser.
 
-.. note:: *SDL_image* has also been added to ports, use it with ``-s USE_SDL_IMAGE=2``. To see a list of all available ports, run ``emcc --show-ports``. For SDL2_image to be useful, you generally need to specify the image formats you are planning on using with -s SDL2_IMAGE_FORMATS='["png"]'. This will also ensure that ``IMG_Init`` works properly. Alternatively, you can use specify ``emcc --use-preload-plugins`` (and ``--preload-file`` your images, so the browser codecs decode them), but then your calls to ``IMG_Init`` will fail.
+.. note:: *SDL_image* has also been added to ports, use it with ``-s USE_SDL_IMAGE=2``. To see a list of all available ports, run ``emcc --show-ports``. For SDL2_image to be useful, you generally need to specify the image formats you are planning on using with e.g. ``-s SDL2_IMAGE_FORMATS='["bmp","png","xpm"]'`` (jpg not in the port). This will also ensure that ``IMG_Init`` works properly. Alternatively, you can use specify ``emcc --use-preload-plugins`` (and ``--preload-file`` your images, so the browser codecs decode them), a patch in the SDL2_image port will load through :c:func:`emscripten_get_preloaded_image_data` but then your calls to ``IMG_Init`` will fail (no built-in support).
```

Comment:
Disclaimer: I more often compile the official SDL2_image version which is fresher and fixes bugs. Still, I had to spend a few hours perusing the emscripten code to understand how this thing about --use-preload-plugins could work and why it wouldn't with the official version.

- jpg: the current SDL2_image port requires a port of jpeg, which isn't readily available through the emscripten ports system (unlike png). It also doesn't try to use the (old) version of jpeg that is bundled in SDL2_image/external/; hence compilation fails with a missing jpeglib.h.

- IMG_Init: can be initialized with [flags](https://www.libsdl.org/projects/SDL_image/docs/SDL_image.html#SEC8) such as `IMG_Init(IMG_INIT_PNG|IMG_INIT_TIF)` to e.g. ensure PNG & TIF support is present in the linked libsdl2_image.so. If you --use-preload-plugins you do not want to compile PNG/BMP/JPEG support but want to rely on the browser (e.g. to optimize size). Consequently that kind of call will fail. Maybe the port's patch could be modified to report support, but that would be only half-true, since that would only work on preloaded resources. From the top of my head, a work-around is not to call IMG_Init or call it with argument 0 / empty flag.

Reply:
When you say `jpg: the current SDL2_image port requires a port of jpeg`, does "current" mean current in emscripten-ports, or in upstream?

Makes sense about IMG_Init - it should report lack of support. We should clarify that more here in this PR, I think. How about something like this?
```
*SDL_image* has also been added to ports, use it with ``-s USE_SDL_IMAGE=2``. To see a list of all available ports, run ``emcc --show-ports``. For SDL2_image to be useful, you generally need to specify the image formats you are planning on using with e.g. ``-s SDL2_IMAGE_FORMATS='["bmp","png","xpm"]'`` (note: jpg support is not in the port yet). This will also ensure that ``IMG_Init`` works properly when you specify those formats. Alternatively, you can use specify ``emcc --use-preload-plugins`` (and ``--preload-file`` your images, so the browser codecs decode them), a code path in the SDL2_image port will load through :c:func:`emscripten_get_preloaded_image_data`, but then your calls to ``IMG_Init`` with those image formats will fail (as while the images will work through preloading, IMG_Init reports no support for those formats, as it doesn't have support compiled in).```

Reply:
SDL2_image (all versions) requires libjpg present.
In its current state, the port lacks an additional port of jpeg.
(What I do instead is emcc-compile both manually without using the emscripten port system.)

I added your changes to the PR branch.

Reply:
I meant:
SDL2_image (all versions) requires libjpg present - _for a JPEG-supporting build of SDL2_image_.

With the current SDL2_image:
```
$ emcc -s USE_SDL_IMAGE=2 -s SDL2_IMAGE_FORMATS='["jpg"]' whatever.c
INFO:root:generating port: sdl2-image-jpg.bc... (this will be cached in "/home/me/.emscripten_cache/asmjs/sdl2-image-jpg.bc" for subsequent builds)
/home/me/.emscripten_ports/sdl2-image/SDL2_image-version_4/IMG_jpg.c:34:10: fatal error: 'jpeglib.h' file not found
#include <jpeglib.h>
         ^~~~~~~~~~~
1 error generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
```

Reply:
Are you saying that currently SDL2_image in emscripten-ports does not support jpg, but newer versions may support it?

Reply:
Hmm, we've been exchanging about this for nearly two weeks so it is clear that I've been unclear :)

The easiest way to clarify all this would be to add support for a new `-s USE_LIBJPEG=1`/`tools/ports/libjpeg.py` - we could add "jpg" to SDL2_IMAGE_FORMATS, no more questions.
Currently that's not the case, not sure if that's a manpower issue, or a wish to keep tools/ports/ minimal?

Reply:
Oh, the missing thing is libjpg itself is not in emscripten-ports? Yeah, that's something we should definitely add, just no one had the time/interest to do it yet.

Ok, I think I get it now, thanks! Merging.